File: windows/Gantry.Windows/Services/ElegooStatusParser.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Gantry.Models;

namespace Gantry.Services;

public static class ElegooStatusParser
{
    public static JsonObject DeepMerge(JsonObject target, JsonObject update)
    {
        var result = (JsonObject)target.DeepClone();
        foreach (var pair in update)
            if (pair.Value is JsonObject next && result[pair.Key] is JsonObject old) result[pair.Key] = DeepMerge(old, next);
            else result[pair.Key] = pair.Value?.DeepClone();
        return result;
    }

    public static PrinterTelemetry Cc2(JsonObject result, PrinterTelemetry? previous = null)
    {
        var value = previous?.Clone() ?? new PrinterTelemetry();
        var machine = result["machine_status"] as JsonObject ?? new();
        var print = result["print_status"] as JsonObject ?? new();
        var extruder = result["extruder"] as JsonObject ?? new();
        var bed = result["heater_bed"] as JsonObject ?? new();
        var chamber = result["ztemperature_sensor"] as JsonObject ?? result["chamber"] as JsonObject ?? new();
        var fans = result["fans"] as JsonObject ?? new();
        var move = result["gcode_move_inf"] as JsonObject ?? result["gcode_move"] as JsonObject ?? new();
        var state = Text(print["state"]).ToLowerInvariant();
        int status = Int(machine["status"]) ?? 1, sub = Int(machine["sub_status"]) ?? -1;
        var errors = machine["exception_status"] as JsonArray;
        if (state == "error" || errors is { Count: > 0 } || status == 14) value.State = PrinterState.Error;
        else if (state.Contains("paus") || sub is 2501 or 2502 or 2505) value.State = PrinterState.Paused;
        else if (state.Contains("complete") || sub == 2077) value.State = PrinterState.Finished;
        else if (state is "printing" or "resuming" || status == 2) value.State = PrinterState.Printing;
        else value.State = PrinterState.Idle;
        if (Int(print["progress"] ?? machine["progress"]) is int progress) value.Progress = Math.Clamp(progress, 0, 100);
        if (Int(print["remaining_time_sec"]) is int seconds) value.RemainingMinutes = Math.Max(0, (int)Math.Round(seconds / 60d));
        if (print.ContainsKey("filename")) value.JobName = EmptyToNull(Text(print["filename"]));
        if (print.ContainsKey("current_layer")) value.CurrentLayer = Int(print["current_layer"]);
        if (print.ContainsKey("total_layer")) value.TotalLayers = Int(print["total_layer"]);
        if (print.ContainsKey("filament_used")) value.FilamentUsedMM = Double(print["filament_used"]);
        if (extruder.ContainsKey("temperature")) value.NozzleTemperature = Double(extruder["temperature"]);
        if (extruder.ContainsKey("target")) value.NozzleTargetTemperature = Double(extruder["target"]);
        if (bed.ContainsKey("temperature")) value.BedTemperature = Double(bed["temperature"]);
        if (bed.ContainsKey("target")) value.BedTargetTemperature = Double(bed["target"]);
        if (chamber.ContainsKey("temperature")) value.ChamberTemperature = Double(chamber["temperature"]);
        value.PartFanPercent = Fan(fans["fan"], value.PartFanPercent);
        value.AuxFanPercent = Fan(fans["aux_fan"], value.AuxFanPercent);
        value.ChamberFanPercent = Fan(fans["box_fan"], value.ChamberFanPercent);
        if (Int(move["speed_mode"]) is int mode)
        {
            value.SpeedLevel = mode + 1;
            value.SpeedPercent = mode switch { 0 => 50, 1 => 100, 2 => 150, 3 => 200, _ => null };
        }
        if (errors is { Count: > 0 })
        {
            value.HmsCodes = errors.Select(Text).ToList();
            value.ErrorCode = (ulong)(Int(errors[0]) ?? 0);
        }
        value.LastUpdated = DateTime.Now;
        return value;
    }

    public static PrinterTelemetry Canvas(JsonObject result, PrinterTelemetry previous)
    {
        var value = previous.Clone();
        if (result["canvas_info"] is not JsonObject info || info["canvas_list"] is not JsonArray canvases) return value;
        int activeCanvas = Int(info["active_canvas_id"]) ?? -1, activeTray = Int(info["active_tray_id"]) ?? -1;
        var groups = new List<FilamentGroup>();
        foreach (var node in canvases.OfType<JsonObject>())
        {
            if (Int(node["connected"]) == 0) continue;
            int canvasId = Int(node["canvas_id"]) ?? 0;
            var trays = (node["tray_list"] as JsonArray)?.OfType<JsonObject>().ToDictionary(x => Int(x["tray_id"]) ?? 0) ?? new();
            var slots = new List<FilamentSlot>();
            for (int trayId = 0; trayId < 4; trayId++)
            {
                var tray = trays.GetValueOrDefault(trayId) ?? new(); bool present = (Int(tray["status"]) ?? 0) > 0;
                var color = EmptyToNull(Text(tray["filament_color"]))?.TrimStart('#'); if (color?.Length == 6) color += "FF";
                slots.Add(new FilamentSlot { Id = $"canvas-{canvasId}-{trayId}", Label = $"{(char)('A' + canvasId)}{trayId + 1}",
                    Material = present ? EmptyToNull(Text(tray["filament_type"] ?? tray["filament_name"])) : null,
                    ColorHex = present ? color : null, RemainingPercent = null,
                    IsActive = present && canvasId == activeCanvas && trayId == activeTray });
            }
            groups.Add(new FilamentGroup { Id = $"canvas-{canvasId}", SourceType = FilamentSourceType.Canvas,
                DisplayName = canvasId == 0 ? "CANVAS" : $"CANVAS {(char)('A' + canvasId)}", DeclaredCapacity = 4, Slots = slots });
        }
        if (groups.Count > 0) { value.FilamentGroups = groups; value.AmsSlots = groups.SelectMany(g => g.LegacyAmsSlots()).ToList(); }
        value.LastUpdated = DateTime.Now; return value;
    }

    public static PrinterTelemetry? Cc1(byte[] data, PrinterTelemetry? previous = null)
    {
        JsonNode? root; try { root = JsonNode.Parse(data); } catch { return null; }
        var status = Find(root, "Status"); if (status is null) return null;
        var value = previous?.Clone() ?? new PrinterTelemetry(); var print = status["PrintInfo"] as JsonObject ?? new();
        int ps = Int(print["Status"]) ?? 0, error = Int(print["ErrorNumber"]) ?? 0;
        value.State = error != 0 || ps == 14 ? PrinterState.Error
            : ps is 5 or 6 ? PrinterState.Paused
            : ps is 1 or 7 or 13 or 15 or 16 ? PrinterState.Printing
            : ps == 9 ? PrinterState.Finished : PrinterState.Idle;
        value.NozzleTemperature = KeepDouble(status, "TempOfNozzle", value.NozzleTemperature);
        value.NozzleTargetTemperature = KeepDouble(status, "TempTargetNozzle", value.NozzleTargetTemperature);
        value.BedTemperature = KeepDouble(status, "TempOfHotbed", value.BedTemperature);
        value.BedTargetTemperature = KeepDouble(status, "TempTargetHotbed", value.BedTargetTemperature);
        value.ChamberTemperature = KeepDouble(status, "TempOfBox", value.ChamberTemperature);
        if (Int(print["Progress"]) is int p) value.Progress = Math.Clamp(p, 0, 100);
        if (print.ContainsKey("CurrentLayer")) value.CurrentLayer = Int(print["CurrentLayer"]);
        if (print.ContainsKey("TotalLayer")) value.TotalLayers = Int(print["TotalLayer"]);
        if (print.ContainsKey("Filename")) value.JobName = EmptyToNull(Text(print["Filename"]));
        if (Double(print["CurrentTicks"]) is double current && Double(print["TotalTicks"]) is double total)
            value.RemainingMinutes = Math.Max(0, (int)Math.Round((total - current) / 60));
        value.ErrorCode = (ulong)error; value.LastUpdated = DateTime.Now; return value;
    }

    private static JsonObject? Find(JsonNode? node, string key)
    {
        if (node is JsonObject obj) { if (obj[key] is JsonObject found) return found; foreach (var child in obj) if (Find(child.Value, key) is { } nested) return nested; }
        else if (node is JsonArray arr) foreach (var child in arr) if (Find(child, key) is { } nested) return nested;
        return null;
    }
    private static int? Int(JsonNode? value) => value is null ? null : int.TryParse(Text(value), out var n) ? n : null;
    private static double? Double(JsonNode? value) => value is null ? null : double.TryParse(Text(value), System.Globalization.NumberStyles.Any, System.Globalization.CultureInfo.InvariantCulture, out var n) ? n : null;
    private static string Text(JsonNode? value) => value is JsonValue scalar && scalar.TryGetValue<string>(out var s) ? s : value?.ToJsonString().Trim('"') ?? "";
    private static string? EmptyToNull(string value) => string.IsNullOrWhiteSpace(value) || value == "null" ? null : value;
    private static int? Fan(JsonNode? node, int? fallback) => node is JsonObject fan && Double(fan["speed"]) is double raw ? Math.Clamp((int)Math.Round(raw / 255 * 100), 0, 100) : fallback;
    private static double? KeepDouble(JsonObject obj, string key, double? fallback) => obj.ContainsKey(key) ? Double(obj[key]) : fallback;
}

File: Sources/Gantry/Services/ElegooStatusParser.swift
import Foundation

enum ElegooStatusParser {
    static func deepMerge(_ base: [String: Any], _ update: [String: Any]) -> [String: Any] {
        var result = base
        for (key, value) in update {
            if let old = result[key] as? [String: Any], let new = value as? [String: Any] {
                result[key] = deepMerge(old, new)
            } else { result[key] = value }
        }
        return result
    }

    static func cc2(result: [String: Any], previous: PrinterTelemetry = PrinterTelemetry()) -> PrinterTelemetry {
        var value = previous
        let machine = result["machine_status"] as? [String: Any] ?? [:]
        let print = result["print_status"] as? [String: Any] ?? [:]
        let extruder = result["extruder"] as? [String: Any] ?? [:]
        let bed = result["heater_bed"] as? [String: Any] ?? [:]
        let chamber = result["ztemperature_sensor"] as? [String: Any] ?? result["chamber"] as? [String: Any] ?? [:]
        let fans = result["fans"] as? [String: Any] ?? [:]
        let move = result["gcode_move_inf"] as? [String: Any] ?? result["gcode_move"] as? [String: Any] ?? [:]
        let internalState = string(print["state"]).lowercased()
        let status = int(machine["status"]) ?? 1
        let substatus = int(machine["sub_status"])
        let errors = machine["exception_status"] as? [Any] ?? []
        if internalState == "error" || !errors.isEmpty || status == 14 { value.state = .error }
        else if [2501, 2502, 2505].contains(substatus ?? -1) || internalState.contains("paus") { value.state = .paused }
        else if substatus == 2077 || internalState.contains("complete") { value.state = .finished }
        else if status == 2 || internalState == "printing" || internalState == "resuming" { value.state = .printing }
        else { value.state = .idle }
        if let progress = int(print["progress"] ?? machine["progress"]) { value.progress = max(0, min(100, progress)) }
        if let seconds = int(print["remaining_time_sec"]) { value.remainingMinutes = max(0, Int((Double(seconds) / 60).rounded())) }
        if print.keys.contains("filename") { value.jobName = optionalString(print["filename"]) }
        if print.keys.contains("current_layer") { value.currentLayer = int(print["current_layer"]) }
        if print.keys.contains("total_layer") { value.totalLayers = int(print["total_layer"]) }
        if print.keys.contains("filament_used") { value.filamentUsedMM = double(print["filament_used"]) }
        if extruder.keys.contains("temperature") { value.nozzleTemperature = double(extruder["temperature"]) }
        if extruder.keys.contains("target") { value.nozzleTargetTemperature = double(extruder["target"]) }
        if bed.keys.contains("temperature") { value.bedTemperature = double(bed["temperature"]) }
        if bed.keys.contains("target") { value.bedTargetTemperature = double(bed["target"]) }
        if chamber.keys.contains("temperature") { value.chamberTemperature = double(chamber["temperature"]) }
        value.partFanPercent = fan(fans["fan"], fallback: value.partFanPercent)
        value.auxFanPercent = fan(fans["aux_fan"], fallback: value.auxFanPercent)
        value.chamberFanPercent = fan(fans["box_fan"], fallback: value.chamberFanPercent)
        if let mode = int(move["speed_mode"]) {
            value.speedLevel = mode + 1
            value.speedPercent = [0: 50, 1: 100, 2: 150, 3: 200][mode]
        }
        if !errors.isEmpty {
            value.hmsCodes = errors.map { string($0) }
            value.errorCode = UInt64(int(errors.first) ?? 0)
        }
        value.lastUpdated = Date()
        return value
    }

    static func canvas(result: [String: Any], previous: PrinterTelemetry) -> PrinterTelemetry {
        guard let info = result["canvas_info"] as? [String: Any],
              let canvases = info["canvas_list"] as? [[String: Any]] else { return previous }
        let activeCanvas = int(info["active_canvas_id"]), activeTray = int(info["active_tray_id"])
        var groups: [FilamentGroup] = []
        for canvas in canvases where int(canvas["connected"]) != 0 {
            let canvasID = int(canvas["canvas_id"]) ?? 0
            let trays = canvas["tray_list"] as? [[String: Any]] ?? []
            let indexed = Dictionary(uniqueKeysWithValues: trays.map { (int($0["tray_id"]) ?? 0, $0) })
            let slots = (0..<4).map { trayID -> FilamentSlot in
                let tray = indexed[trayID] ?? [:]
                let present = (int(tray["status"]) ?? 0) > 0
                var color = optionalString(tray["filament_color"])?.replacingOccurrences(of: "#", with: "")
                if color?.count == 6 { color! += "FF" }
                return FilamentSlot(id: "canvas-\(canvasID)-\(trayID)", label: "\(Character(UnicodeScalar(65 + canvasID)!))\(trayID + 1)",
                    material: present ? optionalString(tray["filament_type"] ?? tray["filament_name"]) : nil,
                    colorHex: present ? color : nil, remainingPercent: nil,
                    isActive: present && canvasID == activeCanvas && trayID == activeTray)
            }
            groups.append(FilamentGroup(id: "canvas-\(canvasID)", sourceType: .canvas,
                displayName: canvasID == 0 ? "CANVAS" : "CANVAS \(Character(UnicodeScalar(65 + canvasID)!))",
                declaredCapacity: 4, humidityPercent: nil, temperatureCelsius: nil, isExternal: false, slots: slots))
        }
        guard !groups.isEmpty else { return previous }
        var value = previous
        value.filamentGroups = groups
        value.amsSlots = groups.flatMap(\.legacyAMSSlots)
        value.lastUpdated = Date()
        return value
    }

    static func cc1(data: Data, previous: PrinterTelemetry = PrinterTelemetry()) -> PrinterTelemetry? {
        guard let root = try? JSONSerialization.jsonObject(with: data), let status = findDictionary(named: "Status", in: root) else { return nil }
        var value = previous
        let print = status["PrintInfo"] as? [String: Any] ?? [:]
        let printStatus = int(print["Status"]) ?? 0
        let error = int(print["ErrorNumber"]) ?? 0
        if error != 0 || printStatus == 14 { value.state = .error }
        else if [5, 6].contains(printStatus) { value.state = .paused }
        else if [1, 7, 13, 15, 16].contains(printStatus) { value.state = .printing }
        else if printStatus == 9 { value.state = .finished }
        else { value.state = .idle }
        for (key, path) in [("TempOfNozzle", \PrinterTelemetry.nozzleTemperature),
                            ("TempTargetNozzle", \PrinterTelemetry.nozzleTargetTemperature),
                            ("TempOfHotbed", \PrinterTelemetry.bedTemperature),
                            ("TempTargetHotbed", \PrinterTelemetry.bedTargetTemperature),
                            ("TempOfBox", \PrinterTelemetry.chamberTemperature)] {
            if status.keys.contains(key) { value[keyPath: path] = double(status[key]) }
        }
        if let progress = int(print["Progress"]) { value.progress = max(0, min(100, progress)) }
        if print.keys.contains("CurrentLayer") { value.currentLayer = int(print["CurrentLayer"]) }
        if print.keys.contains("TotalLayer") { value.totalLayers = int(print["TotalLayer"]) }
        if print.keys.contains("Filename") { value.jobName = optionalString(print["Filename"]) }
        if let current = double(print["CurrentTicks"]), let total = double(print["TotalTicks"]) {
            value.remainingMinutes = max(0, Int(((total - current) / 60).rounded()))
        }
        value.errorCode = UInt64(error); value.lastUpdated = Date()
        return value
    }

    private static func findDictionary(named key: String, in value: Any) -> [String: Any]? {
        if let dictionary = value as? [String: Any] {
            if let found = dictionary[key] as? [String: Any] { return found }
            for child in dictionary.values { if let found = findDictionary(named: key, in: child) { return found } }
        } else if let array = value as? [Any] {
            for child in array { if let found = findDictionary(named: key, in: child) { return found } }
        }
        return nil
    }
    private static func int(_ value: Any?) -> Int? { if let n = value as? NSNumber { return n.intValue }; return Int(string(value)) }
    private static func double(_ value: Any?) -> Double? { if let n = value as? NSNumber { return n.doubleValue }; return Double(string(value)) }
    private static func string(_ value: Any?) -> String { value.map { String(describing: $0) } ?? "" }
    private static func optionalString(_ value: Any?) -> String? { let result = string(value); return result.isEmpty || result == "<null>" ? nil : result }
    private static func fan(_ value: Any?, fallback: Int?) -> Int? {
        guard let object = value as? [String: Any], let raw = double(object["speed"]) else { return fallback }
        return max(0, min(100, Int((raw / 255 * 100).rounded())))
    }
}
